src/ripple/app/misc/Validations.cpp

```diff
@@ -429,25 +429,34 @@ class ValidationsImp : public Validations
         bool anyNew = false;
 
         JLOG (j_.info()) << "Flushing validations";
-        ScopedLockType sl (mLock);
-        for (auto& it: mCurrentValidations)
         {
-            if (it.second)
-                mStaleValidations.push_back (it.second);
-
-            anyNew = true;
-        }
-        mCurrentValidations.clear ();
+            ScopedLockType sl (mLock);
+            for (auto& it: mCurrentValidations)
+            {
+                if (it.second)
+                {
+                    mStaleValidations.push_back (it.second);
+                    anyNew = true;
+                }
+            }
+            mCurrentValidations.clear ();
 
-        if (anyNew)
-            condWrite ();
+            // If there isn't a write in progress already, then write to the
+            // database synchronously.
+            if (anyNew && !mWriting)
+            {
+                mWriting = true;
+                doWrite (sl);
+            }
 
-        while (mWriting)
-        {
-            ScopedUnlockType sul (mLock);
-            std::this_thread::sleep_for (std::chrono::milliseconds (100));
+            // Handle the case where flush() is called while a queuedWrite
+            // is already in progress.
+            while (mWriting)
+            {
+                ScopedUnlockType sul (mLock);
+                std::this_thread::sleep_for (std::chrono::milliseconds (100));
+            }
```

Comment:
That sounds legit to me. The only reason I brought it up was that I noticed yesterday that there are several other instances of `mWriting` being accessed when not under lock. Looking at it again, it appears that the only time it's not explicitly locked is in in `condWrite`, and `condWrite` appears to always be called under lock. 

Soooooooo... Maybe do the same trick in `doWrite` in `condWrite` so that we don't make this same confusing mistake again?